### PR TITLE
Fix several bugs in autolink

### PR DESCRIPTION
**Pull request by rillig**
The main bugs are listed in the second commit. These bugs are fixed in the follow-up commits, as well as some more bugs that I found while fixing the first ones.

### CODE REVIEW

**Comment by M66B** [comment]
What is exactly improved/fixed to the existing auto link?

File: app/src/main/java/eu/faircode/email/HtmlHelper.java [hunk 1]
```diff
@@ -60,7 +60,7 @@
 
 public class HtmlHelper {
     private static final int PREVIEW_SIZE = 250;
-    private static Pattern pattern = Pattern.compile("([http|https]+://[\\w\\S(\\.|:|/)]+)");
+    private static final Pattern pattern = Pattern.compile("https?://[\\w#%&()*+,\\-./:;=?@\\[\\]_|~]*[\\w#)*+\\-/=@\\]|~]");
```
Comment (M66B) [inline app/src/main/java/eu/faircode/email/HtmlHelper.java]: You'll have to explain at least how this regex works / is different from the original regex and why. If you got this regex from somewhere, you'll have to tell where it comes from.

Regex'es like this are notoriously hard to get right, which is why I have kept it as simple as possible. It is just an attempt to auto link 95% of the URLs without failing hard or running into all kind of edge cases.

Comment (rillig) [reply app/src/main/java/eu/faircode/email/HtmlHelper.java]: The new regular expression is partly explained by the newly added unit tests. If you change the regular expression back to its previous version, it will fail some of the tests.

The regular expression consists of 3 parts:
1. `https?://`
2. `[\w#%&()*+,\-./:;=?@\[\]_|~]*`
3. `[\w#)*+\-/=@\]|~]`

Part 1 should be obvious. It's either `http` or `https`.

Part 2 matches any ASCII character that is either allowed in URLs or is commonly found in URLs. That's digits and letters, plus several punctuation characters, ordered by their Unicode code point. Some example characters that are not included here are `<>` and `'"` since these are usually used to separate a URL from the surrounding text. The characters `()` are included though, because when copying a Wikipedia URL, the trailing `)` is not percent-encoded. This last case is debatable. The `)` should probably only be part of the URL if there already is a corresponding `(` in the URL. But then it gets complicated …

Part 3 is a subset of part 2. It is missing characters which usually don't appear at the very end of a URL but are more likely punctuation from the surrounding text. Examples for these are dot, comma, colon.

Comment (M66B) [reply app/src/main/java/eu/faircode/email/HtmlHelper.java]: Matching URIs in a text is problematic and will always have edge cases, so I think I will just switch to Patterns.WEB_URL because it will be *very* hard to do any better.

Source code:
https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/util/Patterns.java

The text replacing can be rewritten by creating Jsoup node elements, which will solve any encoding problems.

Comment (rillig) [reply app/src/main/java/eu/faircode/email/HtmlHelper.java]: Your choice. I just had a look at the test cases from the [PatternsTest](https://android.googlesource.com/platform/frameworks/base/+/master/core/tests/coretests/src/android/util/PatternsTest.java), and they were missing a few that I have in my list. The `WEB_URL` pattern fails at Typo3 and Wikipedia URLs, and it removes a trailing `/` even though it shouldn't.

Comment (M66B) [reply app/src/main/java/eu/faircode/email/HtmlHelper.java]: I am sure there are plenty of other edge cases and that it won't ever be perfect, so choices have to be made. I think Patterns.WEB_URL is as good as it can get.

Comment (M66B) [reply app/src/main/java/eu/faircode/email/HtmlHelper.java]: Thanks for trying to improve this anyway, much appreciated!

**Comment by rillig** [comment]
Whichever regular expression you pick, please make at least sure that the URL explosion from https://github.com/M66B/open-source-email/pull/147/commits/62ac999d1aec2c6c43dba6ed73d708daa3db4602 is properly fixed.

**Comment by M66B** [comment]
Please see the commit for the regex used (Patterns.WEB_URL).
